modules/apps/journal/journal-web/src/main/java/com/liferay/journal/web/internal/info/item/provider/JournalArticleInfoItemFormProvider.java

```diff
@@ -0,0 +1,507 @@
+import com.liferay.info.item.provider.InfoItemFormProvider;
+import com.liferay.info.localized.InfoLocalizedValue;
+import com.liferay.journal.model.JournalArticle;
+import com.liferay.journal.model.JournalArticleDisplay;
+import com.liferay.journal.service.JournalArticleLocalService;
+import com.liferay.journal.util.JournalContent;
+import com.liferay.journal.util.JournalConverter;
+import com.liferay.journal.util.comparator.ArticleVersionComparator;
+import com.liferay.journal.web.internal.asset.JournalArticleDDMFormValuesReader;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.json.JSONUtil;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.service.ServiceContextThreadLocal;
+import com.liferay.portal.kernel.service.UserLocalService;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.FastDateFormatFactoryUtil;
+import com.liferay.portal.kernel.util.LocaleThreadLocal;
+import com.liferay.portal.kernel.util.LocaleUtil;
+import com.liferay.portlet.display.template.PortletDisplayTemplate;
+
+import java.text.Format;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Jürgen Kappler
+ * @author Jorge Ferrer
+ */
+@Component(
+	immediate = true, property = Constants.SERVICE_RANKING + ":Integer=10",
+	service = InfoItemFormProvider.class
+)
+public class JournalArticleInfoItemFormProvider
+	implements InfoItemFormProvider<JournalArticle> {
+
+	@Override
+	public InfoForm getInfoForm() {
+		InfoForm infoForm = new InfoForm(JournalArticle.class.getName());
+
+		infoForm.addAll(_getJournalArticleFields());
+
+		infoForm.add(
+			_infoItemFieldReaderFieldSetProvider.getInfoFieldSet(
+				JournalArticle.class.getName()));
+
+		infoForm.add(
+			_assetEntryInfoItemFieldSetProvider.getInfoFieldSet(
+				AssetEntry.class.getName()));
+
+		infoForm.add(
+			_expandoInfoItemFieldSetProvider.getInfoFieldSet(
+				JournalArticle.class.getName()));
+
+		return infoForm;
+	}
+
+	@Override
+	public InfoForm getInfoForm(JournalArticle article) {
+		DDMStructure ddmStructure = article.getDDMStructure();
+
+		long ddmStructureId = ddmStructure.getStructureId();
+
+		try {
+			return getInfoForm(ddmStructureId);
+		}
+		catch (NoSuchClassTypeException noSuchClassTypeException) {
+			throw new RuntimeException(
+				"Unable to find structure: " + ddmStructureId,
+				noSuchClassTypeException);
+		}
+	}
+
+	@Override
+	public InfoForm getInfoForm(long ddmStructureId)
```

Comment:
Hi @JorgeFerrer,

Why is this a structureId instead of an articleId?

Reply:
First, I have to say that this is an aspect of InfoFormProvider that I'm not satisfied with. I kept what had already been done in InfoDisplayContributor because I tried several alternatives but I wasn't convinced by any of them either. I hope as part of the "Design Critique" we can together find something better.

In any case, to answer your question, what we are identifying here is not the actual article but the subtype (aka classType) of the article, which in the case of JournalArticle is determined by the structure. When this method is used, there isn't necessarily an specific article, but the subtype/classType/structure is known

Makes sense?

Reply:
One more thing, when the article is known, the method to use is getInfoForm(JournalArticle article). If we created a variation of the method that receives an identifier then it would require the implementation to know how to get the object from the identifier (which means less reusability and more code duplication) and also it would only work for a specific way to reference the entity (which doesn't necessarily need to be a long always even if we at Liferay use it in most cases).